File: orchestrator/contracts/holdings.go
package contracts

import (
	"time"
)

// Position represents a single portfolio position holding in HoldingsSnapshot.
type Position struct {
	Ticker         string   `json:"ticker"`
	Quantity       float64  `json:"quantity"`
	AssetClass     string   `json:"asset_class"`
	MarketValueUSD float64  `json:"market_value_usd"`
	AccountType    *string  `json:"account_type,omitempty"` // taxable, retirement
}

// HoldingsSnapshot represents current portfolio holdings.
// Unlike the IPS, this is current-state data — overwritten as holdings change, not versioned/append-only.
type HoldingsSnapshot struct {
	UserID        string     `json:"user_id"`
	AsOf          time.Time  `json:"as_of"`
	Positions     []Position `json:"positions"`
	CashUSD       *float64   `json:"cash_usd,omitempty"`
	TotalValueUSD *float64   `json:"total_value_usd,omitempty"`
}

// Validate returns true if the snapshot meets basic validation rules.
func (h *HoldingsSnapshot) Validate() bool {
	if h.UserID == "" || h.AsOf.IsZero() {
		return false
	}
	for _, pos := range h.Positions {
		if pos.Ticker == "" || pos.Quantity < 0 || pos.AssetClass == "" || pos.MarketValueUSD < 0 {
			return false
		}
	}
	return true
}

File: orchestrator/src/orchestrator/contracts/holdings.py
from datetime import datetime
from enum import Enum
from typing import List, Optional

from pydantic import BaseModel, Field


class AccountType(str, Enum):
    TAXABLE = "taxable"
    RETIREMENT = "retirement"


class Position(BaseModel):
    ticker: str
    quantity: float = Field(ge=0)
    asset_class: str
    market_value_usd: float = Field(ge=0)
    account_type: Optional[AccountType] = None


class HoldingsSnapshot(BaseModel):
    user_id: str
    as_of: datetime
    positions: List[Position]
    cash_usd: Optional[float] = Field(default=None, ge=0)
    total_value_usd: Optional[float] = Field(default=None, ge=0)
